.github/workflows/build.yml

```diff
@@ -148,135 +165,233 @@ jobs:
-          }
           validate-oci-layout temp
 
-          mkdir signed
+          # Login to Docker Hub
+          export DOCKER_CONFIG="$PWD/.docker"
+          mkdir "$DOCKER_CONFIG"
+          trap 'find "$DOCKER_CONFIG" -type f -exec shred -fuvz "{}" + || :; rm -rf "$DOCKER_CONFIG"' EXIT
+          docker login --username "$DOCKER_HUB_USERNAME" --password-stdin <<<"$DOCKER_HUB_PASSWORD"
+          unset DOCKER_HUB_USERNAME DOCKER_HUB_PASSWORD
 
-          args+=(
-            --mount "type=bind,src=$PWD/temp,dst=/doi-build/unsigned" # TODO this currently assumes normalized_builder == "buildkit" and !should_use_docker_buildx_driver -- we need to factor that in later (although this signs the attestations, not the image, so buildkit/buildx is the only builder whose output we *can* sign right now)
-            --mount "type=bind,src=$PWD/signed,dst=/doi-build/signed"
+          # Create signatures
+          dockerArgs+=(
+            --mount "type=bind,src=$PWD/temp,dst=/doi-build/image" # TODO this currently assumes normalized_builder == "buildkit" and !should_use_docker_buildx_driver -- we need to factor that in later (although this signs the attestations, not the image, so buildkit/buildx is the only builder whose output we *can* sign right now)
+            --mount "type=bind,src=$PWD/provenance.json,dst=/doi-build/provenance.json"
+            --mount "type=bind,src=$PWD/.docker,dst=/.docker"
 
             # https://explore.ggcr.dev/?repo=docker/image-signer-verifier
-            docker/image-signer-verifier:0.3.3@sha256:a5351e6495596429bacea85fbf8f41a77ce7237c26c74fd7c3b94c3e6d409c82
-
-            sign
+            "$IMAGE_SIGNER"
+          )
 
-            --envelope-style oci-content-descriptor
+          kmsArg=(
+             # kms key used to sign attestation artifacts
+            --kms="AWS"
+            --kms-region="$AWS_KMS_REGION"
+            --kms-key-ref="$AWS_KMS_KEY_ARN"
 
-            --aws_region "$AWS_KMS_REGION"
-            --aws_arn "awskms:///$AWS_KMS_KEY_ARN"
+            --referrers-dest="$REFERRERS_REPO"  # repo to store attestation artifacts and provenance
+          )
 
-            --input oci:///doi-build/unsigned
-            --output oci:///doi-build/signed
+          # Sign buildkit statements
+          signArgs=(
+            ${kmsArg[@]}
+            --input=oci:///doi-build/image
+            --keep=true  # keep preserves the unsigned attestations generated by buildkit
           )
 
-          docker run "${args[@]}"
+          docker run "${dockerArgs[@]}" sign "${signArgs[@]}"
 
-          validate-oci-layout signed
+          # Attach and sign provenance
+          provArgs=(
+            ${kmsArg[@]}
+            --image=oci:///doi-build/image
+            --statement="/doi-build/provenance.json"
+          )
+          docker run "${dockerArgs[@]}" attest "${provArgs[@]}"
+
+  push:
+    name: Push
+    needs:
+      - build
+      - sign
+      # - verify
+    if: ${{ always() }}
```

Comment:
We don't want this to still run on _failure_, right?  I guess this is how you're managing the issue with the `if:` I noted above? :sob:

Reply:
We want to clear the artifact on failure and there is an extra setting to not clear it for debugging (that's what the of does)